### PR TITLE
[18.05] Strip a spurious Slurm warning from job stderr

### PR DESCRIPTION
**Pull request by nsoranzo**
Reason for the Slurm warning
----------------------------

The Linux kernel memory controller (responsible for the memory cgroup) may run out of cgroup subsystem state (CSS) IDs. When a cgroup is created it is assigned a CSS ID to manage its state. Upon removal of the cgroup the corresponding state information (e.g. cache entries) may still exist and therefore the CSS ID is still held. When multiple frequent short-lived jobs are run on a cluster node, the number of available CSS IDs becomes exhausted and creating a new memory cgroup results in a ENOSPC ("No space left on device"), which is reported by SLURM with the message `unable to add task[pid=<PID>] to memory cg '(null)'` (even if jobs run fine).

There is a bugfix for the kernel that releases the CSS ID upon cgroup destruction, but it appears to be only available after Linux 4.4. We use CentOS 7, which is based on Linux 3.10. The only temporary solutions is to reboot the affected cluster node.

Thanks to @tuxtobin for the detailed analysis above.

Reason for this patch
---------------------

Many tools rely on an empty stderr to determine if the job was successful, either because they were never updated to use `<stdio>`/`detect_errors`, or because the underlying tool returns a non-zero exit code when successful, e.g. `tranalign` from https://toolshed.g2.bx.psu.edu/view/devteam/emboss_5/832c20329690 . Even using a `<regex>` inside `<stdio>` would not work because the Slurm warning contains the word `error`.

### CODE REVIEW

File: lib/galaxy/jobs/runners/slurm.py [hunk 1]
```diff
@@ -149,6 +153,14 @@ def _get_slurm_state():
                     self.work_queue.put((self.fail_job, ajs))
                     return
             if drmaa_state == self.drmaa_job_states.DONE:
+                with open(ajs.error_file, 'r') as rfh:
+                    first_line = rfh.readline()
```
Comment (mvdbeek) [inline lib/galaxy/jobs/runners/slurm.py:157]: We're going to (partially) read in the the error file starting on line 164 again, ~~and that part also contains a way to limit the size of what readline will read (if I understand correctly)~~. Maybe that can be combined / unified ?

Comment (nsoranzo) [reply lib/galaxy/jobs/runners/slurm.py:157]: @mvdbeek That will not read the file from the first line (if `stderr` is larger than 16 MB), and here we need the first line. Also, that block opens the file for read and write, while we need only 'r' here and we must write to a different file.

**Review by natefoo — APPROVED** [review]
This is fantastic, we've recently enabled cgroups in Slurm and started running in to the same issue. I hadn't yet gotten to working around it, looks like now I won't have to. Thanks @nsoranzo!

The Slurm bug is [here](https://bugs.schedmd.com/show_bug.cgi?id=3890).

[RedHat says](https://twitter.com/RedHatSupport/status/1004021678563880960) they've passed my request to backport the fix on to the appropriate team, but for anyone with an RHEL contract, I would encourage you to make a formal request.

**Review by jmchilton — APPROVED** [review]

**Comment by natefoo** [comment]
@nsoranzo lately these messages are appearing for me as:

```
slurmstepd: error: xcgroup_instantiate: unable to create cgroup '/sys/fs/cgroup/memory/slurm/uid_<uid>/job_<id>' : No space left on device
slurmstepd: error: jobacct_gather/cgroup: unable to instanciate job <id> memory cgroup
```

Any chance you're seeing these? I'm running Slurm 17.11.5 (need to upgrade, whoops).

**Comment by nsoranzo** [comment]
@natefoo We aren't, but it seems we're using the even older Slurm 16.05.8 .

**Pull request merged**